[CAP-60] Select Start and Destination

zachariyaJaveri opened the pull request:
### Goal
The purpose of these changes is to enable the ability to choose a starting point and a destination using the UI that was created. It also allows the user to edit their start/destination and will update the path generated automatically.

## Key features
- A new screen that is used for location searching. Currently, it displays a list of the available buildings to choose from and the user must make a selection from among the list. There is also a search field that allows the user to filter the list. This can be expanded in the future to do more generic searches
- Once a user clicks on the search bar (or starting point/destination in the menu under), this new screen opens and allows the user to choose from among the list
- Clicking on the map will also create a route for the user now
- If the user searches for a starting point, the option to use current location is also available

## UI

**Location search page for destination**
![image](https://github.com/user-attachments/assets/e739e842-c1e7-477a-b0ae-e65547e9aba5)

**Location search page for Starting point**
![image](https://github.com/user-attachments/assets/b570f177-ab7d-45a9-a634-40cd8d819186)

## Comments
- There is a good amount of refactoring that was done as part of this pull request in order to follow best software engineering practices
- Since the location search page is new, there is not actually a mockup for it, once one is created, a new work item to update this page will also be created
- There is a menu on the map that display start location and destination, it now pops up when a new destination is selected
- The RecyclerView that is used to display the location items is generic and is provided by default in Android. In the future (when a mockup is created) this will likely need an "item" component of its own

## Issues
- There are no tests that were created for this, these will have to be implemented in the future with another work item since the functionality is desperately needed. Of course, I am not opposed to trying to test this either.
- There is some functionality that I would have liked to refactor into the map class, but that specific piece of code touches the estimated time and it cannot be moved at the moment without a larger refactoring that is not worth it for this pull request

## Food for thought:
- The search bar at the top is used to set a destination, but we have a menu that pops up that effectively offers the same thing, I'm wondering if the search bar should disappear once that menu is opened and reappear when it's closed
- There is currently no way to just "Clear" the start and destination, if there is spare time we can add this
- The menu for showing directions is directly coupled with the maps activity and I can't help but wonder if it should be it's own layout file and we inflate it once a destination is selected (instead of just having it hidden on the page)
- Two line starting locations/destinations hide the estimated time, whoops

zachariyaJaveri committed:
Create location search screen, passed any previous location string and whether it's supposed to be for destination or starting location

zachariyaJaveri committed:
Added filtering search locations

zachariyaJaveri committed:
Location search now returns the selected item, including name, lat and lng

zachariyaJaveri committed:
Modified maps activity to be able to set destination and starting location

zachariyaJaveri committed:
[refactor] Removed dead code, moved some of the responsibility of route parsing to the map class instead of in the activity. Created a function for toggling the transport route instead of creating 4 anonymous functions that do the same thing. Changed some static data to be created using the building manager

zachariyaJaveri committed:
Fixing minor logic error for toggling campuses

zachariyaJaveri committed:
Update UI for location search

zachariyaJaveri committed:
Small change to UI to make it look a bit better

zachariyaJaveri committed:
Make the bottom sheet pop up when a destination is selected

zachariyaJaveri committed:
Added logic for navigation when clicking on the map

zachariyaJaveri committed:
[refactor] Address some concerns from SonarQube

Kayram2710 commented on app/src/main/java/minicap/concordia/campusnav/buildingmanager/ConcordiaBuildingManager.java:
```diff
@@ -85,6 +86,24 @@ public ArrayList<Building> getBuildingsForCampus(CampusName name){
+     * @param buildingName The name of the building you are looking for
+     * @return List of buildings that match or partially match the searched string
+     */
+    public List<Building> searchBuildingsByName(String buildingName) {
+        ArrayList<Building> foundBuildings = new ArrayList<>();
+        String searchName = buildingName.toLowerCase();
+
+        for(Building building: buildings.values()) {
+            if(building.getBuildingName().toLowerCase().contains(searchName)) {
+                foundBuildings.add(building);
+            }
+        }
+
+        return foundBuildings;
```
Im assuming this is getting the names of all the buildings in the ressource bundle, am I correct?

zachariyaJaveri replied:
In a sense, yes. The resource bundle is parsed when the building manager is first created. Then it stores all that information as a collection of objects that can be accessed using the building manager's functions. So if you're looking for a building that's not in our resource bundle then it won't show up

Kayram2710 commented on app/src/main/java/minicap/concordia/campusnav/screens/MapsActivity.java:
```diff
@@ -155,112 +155,167 @@ protected void onCreate(Bundle savedInstanceState) {
-
-        CampusName campusName = null;
-
-        //fill buildings
-        if(!showSGW){
-            campusName = CampusName.SGW;
-        }else{
-            campusName = CampusName.LOYOLA;
-        }
-
-        populateBuildingsList(campusName);
```
There is a problem on the PR that deletes the outlines of Concordia buildings after switching campuses, this might be the reason why, this segment is deleted and not restored anywhere.

zachariyaJaveri replied:
The polygon lines are added when the map is ready. This code was removed because our destination was a dropdown list and this code was responsible for populating that list. I replaced that functionality with an EditText and the new location search screen. On the other hand, I may know why it's actually doing this, when I clear the markers from the map, it might also be clearing the lines. Will fix!

zachariyaJaveri replied:
Fixed in most recent push

Kayram2710 commented on app/src/main/java/minicap/concordia/campusnav/screens/MapsActivity.java:
```diff
@@ -155,112 +155,167 @@ protected void onCreate(Bundle savedInstanceState) {
-            //TODO google maps does not support so for now, travelMode is the same as walking
-            travelMode = "WALK";
-            getUserLocationPath();
+        carButton.setOnClickListener(v -> changeSelectedTravelMethod(carButton, "DRIVE"));
+
+        transitButton.setOnClickListener(v -> changeSelectedTravelMethod(transitButton, "TRANSIT"));
+
+        yourLocationEditText = findViewById(R.id.yourLocationEditText);
+
+        TextInputEditText searchText = findViewById(R.id.genericSearchField);
+
```
This is a very nice system to manipulate the path rendering.

zachariyaJaveri replied:
Thank you!

Kayram2710 commented on app/src/main/java/minicap/concordia/campusnav/map/InternalGoogleMaps.java:
```diff
@@ -97,4 +100,62 @@ public boolean toggleLocationTracking(boolean isEnabled) {
+        mMap.setOnMapClickListener(listener);
+    }
+
+    public void parseRoutePolylineAndDisplay(JSONArray steps) throws JSONException {
```
I belive this class could be heavily tested via Unit testing. Ideally add some for this.

zachariyaJaveri replied:
Some tests have been added

Kayram2710 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -38,6 +38,10 @@
             android:name=".screens.MapsActivity"
             android:exported="false"
             android:label="@string/title_activity_maps" />
+        <activity
+            android:name=".screens.LocationSearchActivity"
+            android:exported="false"
+            android:label="LocationSearch" />
```
For the addition of a new screen I will create a new mockup. If possible reply to this comment letting me know the ambitious extent of this screen, i.e would you want to add anything behind the funtionality in this PR.

zachariyaJaveri replied:
Sounds good! Let me know when it's done and I'll change the UI! In terms of functionality, there are 2 versions of the screen, one for searching for a starting point and one for destination. The only difference is that the starting point page also has a "Current location" button to let the user use their current location

Kayram2710 reviewed (changes requested):
This is a very good addition that adds very well needed technical  features. However there is a few things that need to be addressed:

#### 1. Testing
This PR should have more Unit tests especially for some of the new functions in the InternalGoogleMaps.java class and ConcordiaBuildingManager. 

#### 2. UX 
This addition comes with alot of UI manipulation that definetly concernd me in terms of the UX: 
- The zoom in after every path rendering when clicking path options is very unintuitive, perhaps it would be better to center the camera around the whole path.
- There a few bugs i noticed while testing out the branch, I am not sure if we should merge this to main and report the bugs then or address them now in this PR

###### Bugs found

Concordia building shapes not showing after switching campuses. - I commented as to where i believe this bugs originate from
![Bug](https://media4.giphy.com/media/v1.Y2lkPTc5MGI3NjExdHF0dGowMGp0eDZwb2JsaGtlZnpwYXF0OHY0emNhdHFqaGRjaTU4ZyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/iU5LgqRHGBizQlRPbX/giphy.gif)

Location always starts at Dawson College, i believe this might be intentional but if we were going to merge this to main in this state, we should report it as a bug immidietly, the other issue with this is that we already have a current location tracking implmented and the defaults state dont match.
<img src="https://github.com/user-attachments/assets/5dd348d7-8e6f-4045-b6c3-060a662efaf3" alt="Bug" width="220"/>


In the location tab, there is a button that when clicked just messes up the markers, im assuming this is a placeholder for the confirmation button, however its worth to mention its current state
![Bug](https://media4.giphy.com/media/v1.Y2lkPTc5MGI3NjExN216Z29mZXNoanNqeW4yamtlb3hnemc4emU3N29rYmw0OHM1N283aCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/YW2IdjHLtNzsUMEGYV/giphy.gif)

Kayram2710 commented:
Ideally to merge this PR: fix the outline bug if possible and add two unit tests.

zachariyaJaveri commented:
Your suspicion was correct, the default location is Dawson because if we use the location service it gives the location of the VM. However, upon some research, it seems like we are able to set the location of the phone using the emulator, so I will switch the code and we can start using that instead

zachariyaJaveri committed:
Fixed bug with polylines, now when switching campuses, the polygons remain. Refactored the code for user location to actually use the user's location. Refactored how markers are created, instead of creating the marker options, there are parameterized functions for our needs. Added documentation

zachariyaJaveri commented:
UPDATE:
The button that is messing up markers is not actually a button, it is supposed to be the estimated time, but it is squashed (Essentially you're adding markers because it's clicking on the map under). We will open a bug report for this as I don't think it's worth fixing for this PR

I have addressed the building shapes bug and the extra markers that were appearing when clicking on other places on the map. I have also changed the code to properly use the user's location, we can set the phone's location with the emulator (which I tested for this commit) and it's working properly now

As for centering on the start location when the path is rendered, I think we will open a bug for this, I agree it's not ideal but it will take some changes to how and when the line is displayed and involve adding extra controls. As such, I will leave the code as is for that

Working on tests now.

zachariyaJaveri committed:
Added 4 unit tests for InternalGoogleMaps, centered around testing polylines

zachariyaJaveri committed:
Added 5 more tests for InternalGoogleMaps. The only functionality that cannot be tested in unit tests is Markers because there are fields that are private that cannot be initialized

zachariyaJaveri commented:
I added 9 tests that test the functionality of our InternalGoogleMaps class. I have tested all available method except for addMarker. This function relies on the initialization of private fields that we cannot access. I tried to force the initialization process with mocking but was unsuccessful. I will see if there is another way another time, for now I think these tests are sufficient

zachariyaJaveri committed:
[refactor] Address some small issues from SonarQube

Kayram2710 reviewed (approved):
Everything seems to work, for the unresolved mentioned issues we can report them as bugs and address them when adding further relevent features
